File: frontend/src/api/message.ts
import axios from './axios'

export interface Message {
  id: number
  channel_id: number | null
  conversation_id: number | null
  user_id: number
  content: string
  parent_id: number | null
  is_deleted: boolean
  created_at: string
  updated_at: string | null
}

export interface CreateMessageData {
  content: string
  channel_id: number
  parent_id?: number
}

export interface UpdateMessageData {
  content: string
}

export interface MessageListResponse {
  items: Message[]
  total: number
}

export const messageApi = {
  getChannelMessages: async (
    channelId: number,
    limit: number = 50,
    offset: number = 0
  ): Promise<MessageListResponse> => {
    const response = await axios.get<MessageListResponse>(
      `/api/channels/${channelId}/messages`,
      { params: { limit, offset } }
    )
    return response.data
  },

  createMessage: async (channelId: number, data: CreateMessageData): Promise<Message> => {
    const response = await axios.post<Message>(
      `/api/channels/${channelId}/messages`,
      data
    )
    return response.data
  },

  updateMessage: async (messageId: number, data: UpdateMessageData): Promise<Message> => {
    const response = await axios.put<Message>(`/api/messages/${messageId}`, data)
    return response.data
  },

  deleteMessage: async (messageId: number): Promise<void> => {
    await axios.delete(`/api/messages/${messageId}`)
  },

  getReplies: async (messageId: number): Promise<MessageListResponse> => {
    const response = await axios.get<MessageListResponse>(
      `/api/messages/${messageId}/replies`
    )
    return response.data
  },
}

File: backend/src/schemas/message.py
"""Message schemas."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class MessageCreate(BaseModel):
    """Schema for creating a message."""

    content: str
    channel_id: int | None = None
    conversation_id: int | None = None
    parent_id: int | None = None


class MessageUpdate(BaseModel):
    """Schema for updating a message."""

    content: str


class MessageResponse(BaseModel):
    """Schema for message response."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    channel_id: int | None
    conversation_id: int | None
    user_id: int
    content: str
    parent_id: int | None
    is_deleted: bool
    created_at: datetime
    updated_at: datetime | None


class MessageListResponse(BaseModel):
    """Schema for message list response."""

    items: list[MessageResponse]
    total: int
